lib/shared/bucketing-assembly-script/assembly/eventQueue/eventQueue.ts

```diff
@@ -0,0 +1,168 @@
+        this.aggEventQueue = new Map<string, Map<string, Map<string, i64>>>()
+        return jsonArrFromValueArray(payloads).stringify()
+    }
+
+    onPayloadSuccess(payloadId: string): void {
+        this.requestPayloadManager.markPayloadSuccess(payloadId)
+    }
+
+    onPayloadFailure(payloadId: string, retryable: boolean): void {
+        this.requestPayloadManager.markPayloadFailure(payloadId, retryable)
+    }
+
+    queueEvent(user: DVCPopulatedUser, event: DVCEvent, bucketedConfig: BucketedUserConfig): void {
+        // TODO: Implement max queue size
+        // this.maxEventQueueSize = bucketedConfig?.project.settings.sdkSettings?.eventQueueLimit ?? this.maxEventQueueSize
+
+        const requestEvent = new DVCRequestEvent(event, user.user_id, bucketedConfig.featureVariationMap)
+        this.addEventToQueue(user, requestEvent)
+    }
+
+    private addEventToQueue(user: DVCPopulatedUser, event: DVCRequestEvent): void {
+        // TODO: implement?
+        // if (this.eventQueueSize() >= this.maxEventQueueSize) {
+        //     this.logger.warn(`Max event queue size reached, dropping event: ${event}`)
+        //     this.flushEvents()
+        //     return
+        // }
+
+        const user_id = user.user_id
+        let userEvents: UserEventsBatchRecord
+        if (!this.userEventQueue.has(user_id)) {
+            userEvents = new UserEventsBatchRecord(user, [])
+            this.userEventQueue.set(user_id, userEvents)
+        } else {
+            userEvents = this.userEventQueue.get(user_id)
+            userEvents.user = user
+        }
+        userEvents.events.push(event)
+    }
+
+    queueAggregateEvent(event: DVCEvent, bucketedConfig: BucketedUserConfig): void {
+        this.saveAggEvent(event, bucketedConfig)
+    }
+
+    private saveAggEvent(event: DVCEvent, bucketedConfig: BucketedUserConfig): void {
+        const type = event.type
+        const target = event.target
+        if (!target) {
+            throw new Error('Event missing target to save aggregate event')
+        }
+
+        // TODO: implement?
+        // if (this.eventQueueSize() >= this.maxEventQueueSize) {
+        //     this.logger.warn(`Max event queue size reached, dropping aggregate event: ${event}`)
+        //     this.flushEvents()
+        //     return
+        // }
+
+        let typeAggMap: Map<string, Map<string, i64>>
+        if (this.aggEventQueue.has(type)) {
+            typeAggMap = this.aggEventQueue.get(type)
+        } else {
+            typeAggMap = new Map<string, Map<string, i64>>()
+            this.aggEventQueue.set(type, typeAggMap)
+        }
+
+        let targetAggMap: Map<string, i64>
+        if (typeAggMap.has(target)) {
+            targetAggMap = typeAggMap.get(target)
+        } else {
+            targetAggMap = new Map<string, i64>()
+            typeAggMap.set(target, targetAggMap)
+        }
+
+        // TODO: Do we need to check that the type is aggVariableEvaluated here?
```

Comment:
@andrewtraviss thoughts?

Reply:
yea I guess the only thing is do we want to keep these methods more generic so that we could use them for other events in the future. Maybe I'll add some params to these methods to describe this functionality.